From e6904fbdf7976abbadf3e7e658e9416ca49402f9 Mon Sep 17 00:00:00 2001
From: Vincent Becker <vincentx.becker@intel.com>
Date: Fri, 10 Aug 2012 14:17:33 +0200
Subject: [PATCH] Vibra: Add loading of the vibrator hardware module.

Vibrator HAL is now formed of two hardware modules: one is the default
AOSP and the other is the vendor implementation.

The vendor implementation will be loaded prior to the default one.
For that, the vibrator service has to load the module before
registering itself to the JNI.

Change-Id: I844279f5535289f079d412fdc44c5cb3c9c1130c
Author: Vincent Becker <vincentx.becker@intel.com>
Signed-off-by: Vincent Becker <vincentx.becker@intel.com>
Author-tracking-BZ: 49760 94611
---

diff --git a/services/core/jni/com_android_server_VibratorService.cpp b/services/core/jni/com_android_server_VibratorService.cpp
index 03fbd19..3353a6c 100644
--- a/services/core/jni/com_android_server_VibratorService.cpp
+++ b/services/core/jni/com_android_server_VibratorService.cpp
@@ -22,15 +22,14 @@
 
 #include <utils/misc.h>
 #include <utils/Log.h>
-#include <hardware/vibrator.h>
+#include <hardware/intelvibrator.h>
 
 #include <stdio.h>
 
 namespace android
 {
 
-static hw_module_t *gVibraModule = NULL;
-static vibrator_device_t *gVibraDevice = NULL;
+static struct vibrator_module *gVibraModule = NULL;
 
 static void vibratorInit(JNIEnv /* env */, jobject /* clazz */)
 {
@@ -42,16 +41,12 @@ static void vibratorInit(JNIEnv /* env */, jobject /* clazz */)
 
     if (err) {
         ALOGE("Couldn't load %s module (%s)", VIBRATOR_HARDWARE_MODULE_ID, strerror(-err));
-    } else {
-        if (gVibraModule) {
-            vibrator_open(gVibraModule, &gVibraDevice);
-        }
     }
 }
 
 static jboolean vibratorExists(JNIEnv* /* env */, jobject /* clazz */)
 {
-    if (gVibraModule && gVibraDevice) {
+    if (gVibraModule && gVibraModule->vibrator_exists() > 0) {
         return JNI_TRUE;
     } else {
         return JNI_FALSE;
@@ -60,25 +55,15 @@ static jboolean vibratorExists(JNIEnv* /* env */, jobject /* clazz */)
 
 static void vibratorOn(JNIEnv* /* env */, jobject /* clazz */, jlong timeout_ms)
 {
-    if (gVibraDevice) {
-        int err = gVibraDevice->vibrator_on(gVibraDevice, timeout_ms);
-        if (err != 0) {
-            ALOGE("The hw module failed in vibrator_on: %s", strerror(-err));
-        }
-    } else {
-        ALOGW("Tried to vibrate but there is no vibrator device.");
+    if (gVibraModule) {
+        gVibraModule->vibrator_on(timeout_ms);
     }
 }
 
 static void vibratorOff(JNIEnv* /* env */, jobject /* clazz */)
 {
-    if (gVibraDevice) {
-        int err = gVibraDevice->vibrator_off(gVibraDevice);
-        if (err != 0) {
-            ALOGE("The hw module failed in vibrator_off(): %s", strerror(-err));
-        }
-    } else {
-        ALOGW("Tried to stop vibrating but there is no vibrator device.");
+    if (gVibraModule) {
+        gVibraModule->vibrator_off();
     }
 }
 
